File: src/main/java/org/luxons/sevenwonders/game/boards/BoardElementType.java
package org.luxons.sevenwonders.game.boards;

import java.util.List;

import org.luxons.sevenwonders.game.cards.Color;

public enum BoardElementType {
    CARD {
        @Override
        public int getElementCount(Board board, List<Color> colors) {
            return board.getNbCardsOfColor(colors);
        }
    },
    BUILT_WONDER_STAGES {
        @Override
        public int getElementCount(Board board, List<Color> colors) {
            return board.getWonder().getNbBuiltStages();
        }
    },
    DEFEAT_TOKEN {
        @Override
        public int getElementCount(Board board, List<Color> colors) {
            return board.getMilitary().getNbDefeatTokens();
        }
    };

    public abstract int getElementCount(Board board, List<Color> colors);
}

File: game-engine/src/main/kotlin/org/luxons/sevenwonders/game/boards/BoardElementType.kt
package org.luxons.sevenwonders.game.boards

import org.luxons.sevenwonders.game.cards.Color

enum class BoardElementType {
    CARD {
        override fun getElementCount(board: Board, colors: List<Color>?): Int = board.getNbCardsOfColor(colors!!)
    },
    BUILT_WONDER_STAGES {
        override fun getElementCount(board: Board, colors: List<Color>?): Int = board.wonder.nbBuiltStages
    },
    DEFEAT_TOKEN {
        override fun getElementCount(board: Board, colors: List<Color>?): Int = board.military.nbDefeatTokens
    };

    abstract fun getElementCount(board: Board, colors: List<Color>?): Int
}
